kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ChecksumReader.java

```diff
@@ -0,0 +1,128 @@
+    // First try to load the CRC at given version. If not found or failed to read then try to
+    // find the latest CRC file that is created after the lower bound version or within the last 100
+    // versions if no lower bound is provided.
+    Path crcFilePath = checksumFile(logPath, targetedVersion);
+    Optional<CRCInfo> crcInfoOpt = readChecksumFile(engine, crcFilePath);
+    if (crcInfoOpt.isPresent()
+        ||
+        // we don't expect any more checksum files as it is the first version
+        targetedVersion == 0
+        || targetedVersion == lowerBound) {
+      return crcInfoOpt;
+    }
+    logger.info(
+        "CRC file for version {} not found, attempt to loading version up to {}",
+        targetedVersion,
+        lowerBound);
+
+    Path lowerBoundFilePath = checksumFile(logPath, lowerBound);
+    try (CloseableIterator<FileStatus> crcFiles =
+        engine.getFileSystemClient().listFrom(lowerBoundFilePath.toString())) {
+      List<FileStatus> crcFilesList =
+          toFilteredList(
+              crcFiles,
+              file ->
+                  isChecksumFile(file.getPath())
+                      && checksumVersion(new Path(file.getPath())) <= targetedVersion);
+
+      // pick the last file which is the latest version that has the CRC file
+      if (crcFilesList.isEmpty()) {
+        logger.warn("No checksum files found in the range {} to {}", lowerBound, targetedVersion);
+        return Optional.empty();
+      }
+
+      FileStatus latestCRCFile = crcFilesList.get(crcFilesList.size() - 1);
+      return readChecksumFile(engine, new Path(latestCRCFile.getPath()));
+    } catch (IOException e) {
+      logger.warn("Failed to list checksum files from {}", lowerBoundFilePath, e);
+      return Optional.empty();
+    }
+  }
+
+  private static Optional<CRCInfo> readChecksumFile(Engine engine, Path filePath) {
+    try (CloseableIterator<ColumnarBatch> iter =
+        engine
+            .getJsonHandler()
+            .readJsonFiles(
+                singletonCloseableIterator(FileStatus.of(filePath.toString())),
+                CRCInfo.FULL_SCHEMA,
+                Optional.empty())) {
+      // We do this instead of iterating through the rows or using `getSingularRow` so we
+      // can use the existing fromColumnVector methods in Protocol, Metadata, Format etc
+      if (!iter.hasNext()) {
+        logger.warn("Checksum file is empty: {}", filePath);
+        return Optional.empty();
+      }
+
+      ColumnarBatch batch = iter.next();
+      if (batch.getSize() != 1) {
+        String msg = "Expected exactly one row in the checksum file {}, found {} rows";
+        logger.warn(msg, filePath, batch.getSize());
+        return Optional.empty();
+      }
+
+      long crcVersion = FileNames.checksumVersion(filePath);
+
+      return fromColumnarBatch(engine, crcVersion, batch, 0 /* rowId */, filePath.toString());
```

Comment:
`CRCInfo.fromColumnarBatch`.

`fromColumnarBatch` is used in lots of places in the code --> lets include the classname so it's clear to the reader which class and method you are calling